examples/test/qlib/CsvUtil/csvutil.qtest

```diff
@@ -39,14 +52,64 @@ UK,1234567893,\"Sony, Xperia S\",31052012";
 
         # Iterate through source
         CsvDataIterator i(CsvInput, opts);
-        list l = map $1, i;
-        testAssertion("CsvDataIterator 1", \equalsIterated(), (i, new ListIterator(l)));
+        #list l = map $1, i;
+        testAssertion("CsvDataIterator 1", \equalsIterated(), (new ListIterator(CsvRecords), i));
+
+        CsvStringWriter w( ("write-headers": False, "optimal-quotes" : True, "quote_escape" : '"')+opts );
+        string outstr = w.write(CsvRecords);
+        testAssertion("CsvDataWriter writer", \equals(), (CsvInput+"\n", outstr));  # ending CR is assymetrical when empty lines are not ignored
 
         # test with empty data and header lines
         i = new CsvDataIterator("", ("header-lines": 1));
         testAssertion("CsvDataIterator 2", \i.next(), (), RESULT_FAILURE);
     }
 
+    csvMultiTest() {
+        *hash opts = (
+            "fields": (
+                "type": "int",
+
+                "invoice_no": "string",
+                "date": ("type": "date", "format": "YYYYMMDD"),
+
+                "item_no": "string",
+                "item": "string",
+                "pcs": "int",
+                "price": "number",
+
+                "total": "number",
+            ),
+            "headers": (
+                "header": ("type", "invoice_no", "date"),
+                "item":   ("type", "item_no", "item", "pcs", "price"),
+                "footer": ("type", "total"),
+            ),
+            "resolve_type": softstring sub(hash values) {
+                softint type = exists values{'type'}?values{'type'}:values{'0'};
```

Comment:
thanks David. Is ?? operator in doc? I cannot find it at 
https://qore.org/manual/qore-0.8.6.1/lang/html/operators.html.

-T

Dne 11.5.2016 v 15:43 David Nichols napsal(a):
>
> In examples/test/qlib/CsvUtil/csvutil.qtest 
> <https://github.com/qorelanguage/qore/pull/850#discussion_r62849590>:
>
> > +                "date": ("type": "date", "format": "YYYYMMDD"),
> > +
> > +                "item_no": "string",
> > +                "item": "string",
> > +                "pcs": "int",
> > +                "price": "number",
> > +
> > +                "total": "number",
> > +            ),
> > +            "headers": (
> > +                "header": ("type", "invoice_no", "date"),
> > +                "item":   ("type", "item_no", "item", "pcs", "price"),
> > +                "footer": ("type", "total"),
> > +            ),
> > +            "resolve_type": softstring sub(hash values) {
> > +                softint type = exists values{'type'}?values{'type'}:values{'0'};
>
> this should be:
> softint type = values{'type'} ?? values{'0'};
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/qorelanguage/qore/pull/850/files/396da946e73df1a4d804a809b4757f675b8ba387#r62849590>
>


-- 
Tomas Mandys
Qore Technologies
Email: tomas.mandys@qoretechnologies.com
Tel: +420-604690589
Skype: tmandys@hotmail.com